src/context.cc

```diff
@@ -3957,6 +3957,67 @@ error Context::AddAutotrackerRule(
     return noError;
 }
 
+error Context::UpdateAutotrackerRule(
+    const Poco::Int64 rule_id,
+    const std::string &terms,
+    const Poco::UInt64 pid,
+    const Poco::UInt64 tid) {
+
+    if (!rule_id) {
+        return displayError("missing rule id");
+    }
+    if (terms.empty()) {
+        return displayError("missing terms");
+    }
+    if (!pid && !tid) {
+        return displayError("missing project and task");
+    }
+
+    std::string lowercase = Poco::UTF8::toLower(terms);
+
+    {
+        Poco::Mutex::ScopedLock lock(user_m_);
+        if (!user_) {
+            logger.warning("cannot add autotracker rule, user logged out");
+            return noError;
+        }
+
+        Task* t = nullptr;
+        if (tid) {
+            t = user_->related.TaskByID(tid);
+        }
+        if (tid && !t) {
+            return displayError("task not found");
```

Comment:
Overall, with these errors, I'd probably add a bit more information, especially since it seems they'll get displayed to the user. Getting just "task not found" or "task does not belong to project" when trying to modify an autotracker rule could be quite confusing to the user.

Reply:
User isn't supposed to see these errors. Non-emptiness of the project/task is validated by clients, and the projects get assigned from the list of current user's projects.
I see your point that the error messages can be improved for exceptional situations, but I don't know which use case to optimize messages for, and we have these error messages all over the place, so that would be a separate task IMO.

Reply:
My rationale is that calling `displayError` should happen only to let the user know there's something wrong. These are not errors that just get output to the log but they will bubble through to the UI as well. If the user is actually not supposed to see them, I suggest using `Logger` or returning them to the calling function so the API consumer knows about the mistake they made.

You're right we can't fix them everywhere right now but I think that we should do things right, or at least better, when we're adding new methods to the API.

Reply:
Changed those cases to log the errors instead of showing them to the user, using the existing pattern. You're right, no reason to show them to the user even in exceptional situations.

Reply:
Thanks!